src/rpc/client.cpp

```diff
@@ -6,195 +6,253 @@
 
-CRPCConvertTable::CRPCConvertTable()
-{
-    const unsigned int n_elem =
-        (sizeof(vRPCConvertParams) / sizeof(vRPCConvertParams[0]));
+std::string FormatConversionFailure(const std::string& method, const ConversionFailure& failure) {
+    return examine(failure, match {
+            [&](const UnknownRPCMethod&) {
+                return tinyformat::format("Unknown RPC method, %s", method);
+            },
+            [&](const WrongNumberOfArguments& err) {
+                return tinyformat::format(
+                        "%s for method, %s. Needed between %u and %u, but received %u",
+                        err.providedArgs < err.requiredParams
+                        ? "Not enough arguments"
+                        : "Too many arguments",
+                        method,
+                        err.requiredParams,
+                        err.requiredParams + err.optionalParams,
+                        err.providedArgs);
+            },
+            [](const UnparseableArgument& err) {
+                return std::string("Error parsing JSON:") + err.unparsedArg;
+            }
+        });
+}
 
-    for (unsigned int i = 0; i < n_elem; i++) {
-        members.insert(std::make_pair(vRPCConvertParams[i].methodName,
-                                      vRPCConvertParams[i].paramIdx));
+std::optional<std::pair<std::vector<bool>, std::vector<bool>>>
+ParamsToConvertFor(const std::string& method) {
+    auto search = rpcCvtTable.find(method);
+    if (search != rpcCvtTable.end()) {
+        return search->second;
+    } else {
+        return std::nullopt;
     }
 }
 
-static CRPCConvertTable rpcCvtTable;
-
-/** Non-RFC4627 JSON parser, accepts internal values (such as numbers, true, false, null)
- * as well as objects and arrays.
- */
-UniValue ParseNonRFCJSONValue(const std::string& strVal)
+std::optional<UniValue> ParseNonRFCJSONValue(const std::string& strVal)
 {
     UniValue jVal;
-    if (!jVal.read(std::string("[")+strVal+std::string("]")) ||
-        !jVal.isArray() || jVal.size()!=1)
-        throw runtime_error(string("Error parsing JSON:")+strVal);
-    return jVal[0];
+    if (jVal.read(std::string("[")+strVal+std::string("]")) && jVal.isArray() && jVal.size() == 1) {
+        return jVal[0];
+    } else {
+        return std::nullopt;
+    }
 }
 
-/** Convert strings to command-specific RPC representation */
-UniValue RPCConvertValues(const std::string &strMethod, const std::vector<std::string> &strParams)
+tl::expected<UniValue, ConversionFailure>
+RPCConvertValues(const std::string &method, const std::vector<std::string> &strArgs)
 {
     UniValue params(UniValue::VARR);
+    auto paramsToConvert = ParamsToConvertFor(method);
+    std::vector<bool> requiredParams{}, optionalParams{};
+    if (paramsToConvert.has_value()) {
+        std::tie(requiredParams, optionalParams) = paramsToConvert.value();
+    } else {
+        return tl::expected<UniValue, ConversionFailure>(tl::unexpect, UnknownRPCMethod());
+    }
```

Comment:
How about:
```c++
    if (!paramsToConvert.has_value()) {
        return tl::expected<UniValue, ConversionFailure>(tl::unexpect, UnknownRPCMethod());
    }
    const auto [requiredParams, optionalParams] = paramsToConvert.value();
```

Reply:
Great, thanks!

I also took the opportunity to use the simpler `tl::make_unexpected`.